src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java

```diff
@@ -402,6 +429,145 @@ public void appendMetaColIfNotExist(final String tableName) {
     }
   }
 
+  /**
+   * Check whether the user has the role privilge to do schema evolution and whether the schema
+   * evolution option is enabled on the table
+   *
+   * <p>The permission will be stored in schemaEvolutionPermissionForTables
+   *
+   * @param tableName the name of the table
+   * @param role the role of the user
+   * @return whether schema evolution has the required permission to be performed
+   */
+  @Override
+  public boolean hasSchemaEvolutionPermissionForRole(String tableName, String role) {
+    checkConnection();
+    InternalUtils.assertNotEmpty("tableName", tableName);
+    String query = "show grants on table identifier(?)";
+    ResultSet result = null;
+    boolean hasRolePrivilege = false;
+    boolean hasTableOptionEnabled = true;
+    String myRole = SchematizationUtils.formatRoleName(role);
+    // the schema evolution permission is still in PrPr, so it is left as true here
+    try {
+      PreparedStatement stmt = conn.prepareStatement(query);
+      stmt.setString(1, tableName);
+      result = stmt.executeQuery();
+      while (result.next()) {
+        logInfo(String.format("Role: %s, Privilege: %s", result.getString(6), result.getString(2)));
+        if (!result.getString(6).equals(myRole)) {
+          continue;
+        }
+        if (result.getString(2).equals("EVOLVE SCHEMA")
+            || result.getString(2).equals("ALL")
+            || result.getString(2).equals("OWNERSHIP")) {
+          hasRolePrivilege = true;
+        }
+      }
+      stmt.close();
+    } catch (SQLException e) {
+      hasRolePrivilege = true;
+      // if the table doesn't exist, we will create it, and thus we will have the privilege
+    }
+    this.schemaEvolutionPermissionForTables.put(
+        tableName, hasRolePrivilege && hasTableOptionEnabled);
+    return hasRolePrivilege && hasTableOptionEnabled;
+  }
+
+  /**
+   * Check if the role used by the connector has the permission to do schema evolution on the table
+   * through a cached map
+   *
+   * @param tableName the name of the table
+   * @return whether we have the permission to do schema evolution on the table
+   */
+  @Override
+  public boolean getSchemaEvolutionPermission(String tableName) {
```

Comment:
```suggestion
  public boolean hasSchemaEvolutionPermission(String tableName) {
```

Reply:
And is this cached? What if the permission is updated?

Reply:
I guess the costumer will have to restart the connector. It makes no sense to query the permission every time we failed to send the records.